File: POTD/Mathematical/Geometry/Python/1861. Rotating the Box.py

# SIUIING BRUTE FORCE APPROACH
class Solution:
    def rotateTheBox(self, boxGrid: List[List[str]]) -> List[List[str]]:
        m=len(boxGrid)
        n=len(boxGrid[0])

        res=[['.' for _ in range(n)] for _ in range(m)]

        for i in range(m):
            stones=0
            for j in range(n):
                if boxGrid[i][j]=='#':
                    stones+=1
                
                if(boxGrid[i][j]=='*'):
                    res[i][j]='*'
                    col=j-1

                    while stones>0 and col>=0:
                        res[i][col]='#'
                        col-=1
                        stones-=1
            
            col=n-1
            while stones>0 and col>=0:
                res[i][col]='#'
                col-=1
                stones-=1
            
        
        rotatedRes=[['.' for _ in range(m)] for _ in range(n)]

        for i in range(m):
            for j in range(n):
                rotatedRes[j][m-i-1]=res[i][j]
        
        return rotatedRes

File: POTD/Mathematical/Geometry/Cpp/1861. Rotating the Box.cpp

// SIUUING BRUTE FORCE APPROACH
class Solution {
public:
    vector<vector<char>> rotateTheBox(vector<vector<char>>& boxGrid) {
        int m=boxGrid.size();
        int n=boxGrid[0].size();

        vector<vector<char>> res(m,vector<char>(n,'.'));

        for(int i=0;i<m;i++){
            int stones=0;
            for(int j=0;j<n;j++){
                if(boxGrid[i][j]=='#') stones++;

                if(boxGrid[i][j]=='*'){
                    // populate 
                    res[i][j]='*';
                    int col=j-1;
                    while(stones>0 && col>=0){
                        res[i][col]='#';
                        col--;
                        stones--;
                    }
                }
            }


            // checking if we have some stones left
            int col=n-1;
            while(stones>0 && col>=0){
                res[i][col]='#';
                col--;
                stones--;
            }
        }


        // rotate it by 90 degrees
        vector<vector<char>> rotatedRes(n,vector<char>(m));

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                rotatedRes[j][m-i-1]=res[i][j];
            }
        }

        return rotatedRes;
    }
};
